### PR TITLE
Optimizer: avoid every rule must recursive children in optimizer

### LINKED ISSUE #4613: rules don't need to recursion inside themselves
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4613 .

# Rationale for this change

Currently, we must do optimize recursively in each rules, which is troubled.

It make it's a little hard for putting some rule into optimizer.

# What changes are included in this PR?

avoid every rule must recursive children in optimizer

And I modify the `eliminate_limit` as an example.

When `apply_order` is None, it's completely compatible with original behavior.

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by jackwener** [comment]
wait for #4599 merge.

File: datafusion/optimizer/src/eliminate_limit.rs [hunk 1]
```diff
@@ -40,37 +42,59 @@ impl OptimizerRule for EliminateLimit {
...
-                    }
+            }
+            None => {
+                if limit.skip == 0 {
+                    let input = limit.input.as_ref();
+                    // input also can be Limit, so we should apply again.
+                    return Ok(Some(
+                        self.try_optimize(input, _optimizer_config)?
+                            .unwrap_or_else(|| input.clone()),
+                    ));
```
Comment (jackwener) [inline datafusion/optimizer/src/eliminate_limit.rs:78]: 🎉we don't need to recurse optimize children inside rule.

**Comment by jackwener** [comment]
This is one thing that I wanted to do half a year ago🚀 .
It's done now🎉.

File: datafusion/optimizer/src/optimizer.rs [hunk 2]
```diff
@@ -69,6 +69,11 @@ pub trait OptimizerRule {
 
     /// A human readable name for this optimizer rule
     fn name(&self) -> &str;
+
+    /// If a rule use default None, its should traverse recursively plan inside itself
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        None
+    }
```
Comment (jackwener) [inline datafusion/optimizer/src/optimizer.rs:76]: It' compatible with origin

**Comment by jackwener** [comment]
PTAL @alamb @Dandandan @liukun4515

**Comment by alamb** [comment]
I plan to look at this PR carefully tomorrow https://github.com/influxdata/influxdb_iox/pull/6407

**Comment by tustvold** [comment]
We definitely shouldn't hold this up on a theoretical future change, but just FYI - #4627

**Comment by jackwener** [comment]
has resolved conflict, now there are clear

**Comment by jackwener** [comment]
> We definitely shouldn't hold this up on a theoretical future change, but just FYI - #4627

A nice proposal, I have try to do some thing to improve it, but I didn't find a suitable method.

**Comment by mingmwang** [comment]
@jackwener I think you can leverage the `TreeNodeRewritable` trait to unify the code. 
And there are two downsides of the approach:
1) There will be additional `clone()` call to the plan structs. 
2) When implement rules, if we want to implement the rule as a top down process, need to be very careful with transform_down() closures, it might cause stack over flow(because of no duplication invocation check).

**Comment by mingmwang** [comment]
BTW, some complex rules might be a mixed up of bottom-up and top-down processes.

**Comment by jackwener** [comment]
Good advice for me, thanks @mingmwang

File: datafusion/optimizer/src/optimizer.rs [hunk 3]
```diff
@@ -274,6 +284,77 @@ impl Optimizer {
         debug!("Optimizer took {} ms", start_time.elapsed().as_millis());
         Ok(new_plan)
     }
+
+    fn optimize_node(
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:288]: What value does this function add? In other words, why not call `rule.try_optimize` directly at the callsite 🤔

File: datafusion/optimizer/src/optimizer.rs [hunk 4]
```diff
@@ -274,6 +284,77 @@ impl Optimizer {
...
+        &self,
+        rule: &Arc<dyn OptimizerRule + Send + Sync>,
+        plan: &LogicalPlan,
+        optimizer_config: &mut OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        match rule.apply_order() {
+            Some(order) => match order {
+                ApplyOrder::TopDown => {
+                    let optimize_self_opt =
+                        self.optimize_node(rule, plan, optimizer_config)?;
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:334]: I took a shot at rewriting this stuff to use Result::and_then, etc and I think how you have written it here is clearer 👍

File: datafusion/optimizer/src/optimizer.rs (hunk 3 above)
Comment (jackwener) [reply datafusion/optimizer/src/optimizer.rs:288]: I forgot to add future TODO:
we can do batch optimize
```rust
for rule in rules:
    rule.try_optimize(rule)
```

File: datafusion/optimizer/src/optimizer.rs [hunk 5]
```diff
@@ -274,6 +284,77 @@ impl Optimizer {
...
+            .map(|(i, o)| match o {
+                Some(plan) => plan,
+                None => (*(inputs.get(i).unwrap())).clone(),
+            })
+            .collect::<Vec<_>>();
+
+        Ok(Some(plan.with_new_inputs(new_inputs.as_slice())?))
+    }
+
+    pub fn optimize_recursively(
```
[the remark this answers is not in the document]
Comment (jackwener) [reply datafusion/optimizer/src/optimizer.rs:362]: Sometime we need use it in UT😂

**Comment by jackwener** [comment]
Thanks @alamb review!❤️ Advices is used for me.

**Comment by jackwener** [comment]
CI fail due to clippy, fix in #4652

**Comment by jackwener** [comment]
> CI fail due to clippy, fix in #4652

fix it, and rebase.

File: datafusion/optimizer/src/optimizer.rs [hunk 6]
```diff
@@ -147,6 +154,44 @@ pub struct Optimizer {
     pub rules: Vec<Arc<dyn OptimizerRule + Send + Sync>>,
 }
 
+/// If a rule is with `ApplyOrder`, it means the optimizer will derive to handle children instead of
+/// recursively handling in rule.
+/// We just need handle a subtree pattern itself.
+///
+/// Notice: **sometime** result after optimize still can be optimized, we need apply again.
+///
+/// Usage Example: Merge Limit (subtree pattern is: Limit-Limit)
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:163]: Thank you

**Comment by alamb** [comment]
This PR had conflicts, so I took the liberty of resolving them and pushing a fix as @jackwener had already done so 👍 

Once CI passes I plan to merge this in again. Thanks again @jackwener

**Pull request merged**